Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp

```diff
@@ -263,4 +265,97 @@ void RenderTreeUpdater::GeneratedContent::removeAfterPseudoElement(Element& elem
+            m_updater.m_builder.destroy(*writingSuggestionsRenderer);
+    };
+
+    if (!renderer.canHaveChildren())
+        return;
+
+    if (!renderer.element())
+        return;
+
+    auto& editor = renderer.element()->document().editor();
+
+    auto isWritingSuggestionElement = renderer.element() == editor.writingSuggestionsContainerElement();
+    if (!isWritingSuggestionElement)
+        return;
+
+    auto* writingSuggestionData = editor.writingSuggestionData();
+    if (!writingSuggestionData) {
+        destroyWritingSuggestionsIfNeeded();
+        return;
+    }
+
+    auto style = renderer.getCachedPseudoStyle({ PseudoId::InternalWritingSuggestions }, &renderer.style());
+    if (!style || style->display() == DisplayType::None) {
+        destroyWritingSuggestionsIfNeeded();
+        return;
+    }
+
+    auto* firstChildText = dynamicDowncast<RenderText>(renderer.firstChild());
+    if (!firstChildText) {
+        destroyWritingSuggestionsIfNeeded();
+        return;
+    }
+
+    auto textWithoutSuggestion = firstChildText->text();
+
+    auto offset = writingSuggestionData->offset();
+    auto prefix = textWithoutSuggestion.substring(0, offset);
+    auto suffix = textWithoutSuggestion.substring(offset);
+
+    firstChildText->setText(prefix);
+
+    auto newStyle = RenderStyle::clone(*style);
+    if (auto writingSuggestionsRenderer = editor.writingSuggestionRenderer()) {
+        writingSuggestionsRenderer->setStyle(WTFMove(newStyle));
+
+        auto* writingSuggestionsText = dynamicDowncast<RenderText>(writingSuggestionsRenderer->firstChild());
+        if (!writingSuggestionsText) {
+            ASSERT_NOT_REACHED();
+            destroyWritingSuggestionsIfNeeded();
+            return;
+        }
+
+        writingSuggestionsText->setText(writingSuggestionData->content());
+
+        auto* suffixText = dynamicDowncast<RenderText>(writingSuggestionsRenderer->nextSibling());
+        if (!suffixText) {
+            ASSERT_NOT_REACHED();
+            destroyWritingSuggestionsIfNeeded();
+            return;
+        }
+
+        suffixText->setText(suffix);
+    } else {
+        auto newWritingSuggestionsRenderer = WebCore::createRenderer<RenderInline>(RenderObject::Type::Inline, renderer.document(), WTFMove(newStyle));
+        newWritingSuggestionsRenderer->initializeStyle();
+
+        auto writingSuggestionsText = WebCore::createRenderer<RenderText>(RenderObject::Type::Text, renderer.document(), writingSuggestionData->content());
+        m_updater.m_builder.attach(*newWritingSuggestionsRenderer, WTFMove(writingSuggestionsText));
+
+        editor.setWritingSuggestionRenderer(*newWritingSuggestionsRenderer.get());
+        m_updater.m_builder.attach(renderer, WTFMove(newWritingSuggestionsRenderer));
+
+        auto* prefixNode = firstChildText->textNode();
+        if (!prefixNode) {
+            ASSERT_NOT_REACHED();
+            destroyWritingSuggestionsIfNeeded();
+            return;
+        }
```

Comment:
I still have trouble understanding the implications of this. The text in DOM and rendering side will be out of sync so things like text iterators will return surprising results that include the suggestion text. It seems this could still easily break things. I'm going to say r+ to the rendering/style parts to make progress, trusting you will address any issues that will raise. You'll need to find another reviewer for the editing side.

Reply:
I also have trouble understanding why the empty placeholder with shadow DOM approach was never entertained.

Reply:
It'd be great to see some test coverage on this out-of-sync part of the tree getting mutated by JS (i.e. content editing and JS mutating the same time)

Reply:
I am actively working on write a bunch of test sites, and will create layout tests from them in a follow up PR.